internal/hcs/process.go

```diff
@@ -361,46 +375,60 @@ func (process *Process) Stdio() (stdin io.Writer, stdout, stderr io.Reader) {
 
 // CloseStdin closes the write side of the stdin pipe so that the process is
 // notified on the read side that there is no more data in stdin.
-func (process *Process) CloseStdin(ctx context.Context) error {
+func (process *Process) CloseStdin(ctx context.Context) (err error) {
+	operation := "hcs::Process::CloseStdin"
+	ctx, span := trace.StartSpan(ctx, operation) //nolint:ineffassign,staticcheck
+	defer span.End()
+	defer func() { oc.SetSpanStatus(span, err) }()
+	span.AddAttributes(
+		trace.StringAttribute("cid", process.SystemID()),
+		trace.Int64Attribute("pid", int64(process.processID)))
+
 	process.handleLock.RLock()
 	defer process.handleLock.RUnlock()
 
-	operation := "hcs::Process::CloseStdin"
-
 	if process.handle == 0 {
 		return makeProcessError(process, operation, ErrAlreadyClosed, nil)
 	}
 
-	modifyRequest := processModifyRequest{
-		Operation: modifyCloseHandle,
-		CloseHandle: &closeHandle{
-			Handle: stdIn,
-		},
+	process.stdioLock.Lock()
+	defer process.stdioLock.Unlock()
+	if process.stdin == nil {
+		return nil
 	}
 
-	modifyRequestb, err := json.Marshal(modifyRequest)
-	if err != nil {
-		return err
-	}
+	// The HcsModifyProcess request to close stdin will fail if the process has already
+	// exited or been closed before.
+	if !process.exited() {
+		modifyRequest := processModifyRequest{
+			Operation: modifyCloseHandle,
+			CloseHandle: &closeHandle{
+				Handle: stdIn,
+			},
+		}
 
-	resultJSON, err := vmcompute.HcsModifyProcess(ctx, process.handle, string(modifyRequestb))
-	events := processHcsResult(ctx, resultJSON)
-	if err != nil {
-		return makeProcessError(process, operation, err, events)
+		var b []byte
+		b, err = json.Marshal(modifyRequest)
+		// don't return on errors, and still try to close the stream from the host
+		if err == nil {
+			var resultJSON string
+			resultJSON, err = vmcompute.HcsModifyProcess(ctx, process.handle, string(b))
+			events := processHcsResult(ctx, resultJSON)
+			if err != nil {
+				err = makeProcessError(process, operation, err, events)
+			}
+		}
```

Comment:
Is this just so we always call process.stdin.Close() instead of early returning if the marshal fails?

Reply:
yeah, it feels weird to give up on closing stdin on the host side just because the marshaling failed. Especially since the marshalling is an internal detail and the error is salvageable.